lib/billy/ssl/authority.rb

```diff
@@ -0,0 +1,123 @@
+
+module Billy
+  # This class is dedicated to the generation of a brand new certificate
+  # authority which can be picked up by a browser to verify and secure any
+  # communication with puffing billy. This authority certificate will be
+  # generated once on runtime and will sign each request certificate. So
+  # we do not have to deal with outdated certificates or stuff like that.
+  #
+  # The resulting certificate authority is at its bare minimum to keep
+  # things simple and snappy. We do not handle a certificate revoke list
+  # (CRL) nor any other special key handling, even if we enable these
+  # extensions. It's just a mimic of the mighty mitmproxy certificate
+  # authority file.
+  class Authority
+    attr_reader :key, :cert
+
+    # The authority generation does not require any arguments from outside
+    # of this class definition. We just generate the certificate and thats
+    # it.
+    #
+    # Example:
+    #
+    #   ca = Billy::Authority.new
+    #   [ca.cert_file, ca.key_file]
+    def initialize
+      @key = OpenSSL::PKey::RSA.new(2048)
+      @cert = generate
+    end
+
+    # Write out the private key to file (PEM format) and give back the
+    # file path. This will produce a temporary file which will be remove
+    # after the current process terminates.
+    def key_file
+      path = File.join(Billy.config.certs_path, 'ca.key')
+      FileUtils.mkdir_p(File.dirname(path))
+      File.write(path, key.to_pem)
+      path
+    end
+
+    # Write out the certifcate to file (PEM format) and give back the
+    # file path. This will produce a temporary file which will be remove
+    # after the current process terminates.
+    def cert_file
+      path = File.join(Billy.config.certs_path, 'ca.crt')
+      FileUtils.mkdir_p(File.dirname(path))
+      File.write(path, cert.to_pem)
+      path
+    end
+
+    private
+
+    # Defines a static list of available extensions on the certificate.
+    def extensions
+      [
+        # ln_sn, value, critical
+        ['basicConstraints', 'CA:TRUE', true],
+        ['keyUsage', 'keyCertSign, cRLSign', true],
+        ['subjectKeyIdentifier', 'hash', false],
+        ['authorityKeyIdentifier', 'keyid:always', false]
+      ]
+    end
+
+    # Give back the static subject name of the certificate.
+    def name
+      ['CN=Puffing Billy', 'O=Puffing Billy'].join('/')
+                                             .prepend('/')
+                                             .concat('/')
+    end
+
+    # Give back an appropriate date for the beginning of this
+    # certificate life. We give back now 2 days ago.
+    def valid_from
+      Time.now - (2 * 24 * 60 * 60)
+    end
+
+    # Give back an appropriate date for the end of this certificate life.
+    # We give back now in 2 days.
+    def valid_to
+      Time.now + (2 * 24 * 60 * 60)
+    end
+
+    # Generate a random serial number for the certificate.
+    def serial
+      Time.now.to_i + rand(100_000_000_000)
```

Comment:
How does `Time.now.to_i` help here?  We're still getting a random number with just the `rand` half.

Reply:
`rand` should be good enough. The serial should be unique for each cert with the same subject. Otherwise, a browser could cache it, which is highly undesired.